### PR TITLE
qdisk: account for overwriting in get_empty_space()

### PR DESCRIPTION
**Pull request by alltilla**
No news file needed.

Related PR #4356.

Signed-off-by: Attila Szakacs <attila.szakacs@axoflow.com>

### CODE REVIEW

File: modules/diskq/qdisk.c [hunk 1]
```diff
@@ -408,14 +408,107 @@ qdisk_get_empty_space(QDisk *self)
 {
   gint64 wpos = qdisk_get_writer_head(self);
   gint64 bpos = qdisk_get_backlog_head(self);
+  gint64 disk_buf_size = qdisk_get_maximum_size(self);
 
-  if (wpos > bpos)
+  if (wpos < disk_buf_size && bpos < disk_buf_size)
```
Comment (MrAnno) [inline modules/diskq/qdisk.c:418]: Could you write a (unit) test for this?

Comment (alltilla) [reply modules/diskq/qdisk.c:418]: Sure!

File: modules/diskq/qdisk.c [hunk 2]
```diff
@@ -408,14 +408,107 @@ qdisk_get_empty_space(QDisk *self)
...
+          // or
+          //
+          // 0   RESERVED          DBS=B=W       FS
+          // |---|------ ... ------|-------------|
+          //      ^^^^^^^^^^^^^^^^^
+          g_assert(self->hdr->length == 0);
+          return disk_buf_size - QDISK_RESERVED_SPACE;
+        }
+    }
+  else
```
Comment (MrAnno) [inline modules/diskq/qdisk.c:484]: The inner ifs are easy to follow, but this else branch might need some clean-code trick or other explanation:
This branch is about when the write and backlog head are on two different sides of DBS.

**Review by MrAnno — DISMISSED** [review]
LGTM otherwise.

File: modules/diskq/qdisk.c (hunk 1 above)
Comment (alltilla) [reply modules/diskq/qdisk.c:418]: Done

File: modules/diskq/qdisk.c (hunk 2 above)
Comment (alltilla) [reply modules/diskq/qdisk.c]: I think I have made this a bit more clearer.

**Comment by MrAnno** [comment]
@kira-syslogng do stresstest

**Comment by kira-syslogng** [comment]
Kira-stress-test: Build SUCCESS

**Pull request merged**